src/core/labeling/rules/qgslabelingenginerule_impl.cpp

```diff
@@ -0,0 +1,632 @@
+      if ( GEOSPreparedDistanceWithin_r( geosctxt, candidateGeos, featureCandidate.get(), mDistanceMapUnits ) )
+      {
+        return mMustBeDistant;
+      }
+#else
+      QgsDebugError( QStringLiteral( "This rule requires GEOS 3.10+" ) );
+      return false;
+#endif
+    }
+    catch ( GEOSException &e )
+    {
+      QgsDebugError( QStringLiteral( "GEOS exception: %1" ).arg( e.what() ) );
+    }
+  }
+
+  return !mMustBeDistant;
+}
+
+//
+// QgsLabelingEngineRuleMinimumDistanceLabelToFeature
+//
+
+QgsLabelingEngineRuleMinimumDistanceLabelToFeature::QgsLabelingEngineRuleMinimumDistanceLabelToFeature() = default;
+QgsLabelingEngineRuleMinimumDistanceLabelToFeature::~QgsLabelingEngineRuleMinimumDistanceLabelToFeature() = default;
+
+QgsLabelingEngineRuleMinimumDistanceLabelToFeature *QgsLabelingEngineRuleMinimumDistanceLabelToFeature::clone() const
+{
+  std::unique_ptr< QgsLabelingEngineRuleMinimumDistanceLabelToFeature> res = std::make_unique< QgsLabelingEngineRuleMinimumDistanceLabelToFeature >();
+  copyCommonProperties( res.get() );
+  return res.release();
+}
+
+QString QgsLabelingEngineRuleMinimumDistanceLabelToFeature::id() const
+{
+  return QStringLiteral( "minimumDistanceLabelToFeature" );
+}
+
+QString QgsLabelingEngineRuleMinimumDistanceLabelToFeature::displayType() const
+{
+  return QObject::tr( "Push Labels Away from Features" );
+}
+
+QString QgsLabelingEngineRuleMinimumDistanceLabelToFeature::description() const
+{
+  QString res = QStringLiteral( "<b>%1</b>" ).arg( name().isEmpty() ? displayType() : name() );
+  if ( labeledLayer() && targetLayer() )
+  {
+    res += QStringLiteral( "<p>" ) + QObject::tr( "Labels from <i>%1</i> must be at least %2 %3 from features in <i>%4</i>" ).arg(
+             labeledLayer()->name(),
+             QString::number( distance() ),
+             QgsUnitTypes::toAbbreviatedString( distanceUnit() ),
+             targetLayer()->name()
+           ) + QStringLiteral( "</p>" );
+  }
+  return res;
+}
+
+
+//
+// QgsLabelingEngineRuleMaximumDistanceLabelToFeature
+//
+
+QgsLabelingEngineRuleMaximumDistanceLabelToFeature::QgsLabelingEngineRuleMaximumDistanceLabelToFeature()
+{
+  mMustBeDistant = false;
+}
+
+QgsLabelingEngineRuleMaximumDistanceLabelToFeature::~QgsLabelingEngineRuleMaximumDistanceLabelToFeature() = default;
+
+QgsLabelingEngineRuleMaximumDistanceLabelToFeature *QgsLabelingEngineRuleMaximumDistanceLabelToFeature::clone() const
+{
+  std::unique_ptr< QgsLabelingEngineRuleMaximumDistanceLabelToFeature > res = std::make_unique< QgsLabelingEngineRuleMaximumDistanceLabelToFeature >();
+  copyCommonProperties( res.get() );
+  return res.release();
+}
+
+QString QgsLabelingEngineRuleMaximumDistanceLabelToFeature::id() const
+{
+  return QStringLiteral( "maximumDistanceLabelToFeature" );
+}
+
+QString QgsLabelingEngineRuleMaximumDistanceLabelToFeature::displayType() const
+{
+  return QObject::tr( "Pull Labels toward Features" );
```

Comment:
I'm pretty sure one must capitalize toward(s) in a title case sentence.

```suggestion
  return QObject::tr( "Pull Labels Toward Features" );
```